core/src/main/java/com/linecorp/armeria/client/DefaultDnsQueryLifecycleObserver.java

```diff
@@ -0,0 +1,110 @@
+
+    private final PrometheusMeterRegistry registry;
+    private final MeterIdPrefix meterIdPrefix;
+    private final DnsQuestion question;
+    private static final String NAME_TAG = "name";
+    private static final String RESULT_TAG = "result";
+    private static final String SERVER_TAG = "server";
+    private static final String CODE_TAG = "code";
+    private static final String CAUSE_TAG = "cause";
+    private static final String CNAME_TAG = "cname";
+
+    /**
+     * Accepts meterRegistry.
+     * @param meterRegistry {@link PrometheusMeterRegistry} PrometheusMeterRegistry to capture metrics.
+     * @param question {@link DnsQuestion} DnsQuestion.
+     */
+    DefaultDnsQueryLifecycleObserver(PrometheusMeterRegistry meterRegistry,
+                                     DnsQuestion question, MeterIdPrefix prefix) {
+        registry = meterRegistry;
+        meterIdPrefix = prefix;
+        this.question = question;
+    }
+
+    @Override
+    public void queryWritten(InetSocketAddress dnsServerAddress, ChannelFuture future) {
+        registry.counter(meterIdPrefix.name().concat(".written"),
+                Arrays.asList(Tag.of(NAME_TAG, question.name()),
+                        Tag.of(SERVER_TAG, dnsServerAddress.getAddress().toString()))).increment();
+    }
+
+    @Override
+    public void queryCancelled(int queriesRemaining) {
+        registry.counter(meterIdPrefix.name().concat(".cancelled"),
+                NAME_TAG, question.name()).increment();
+    }
+
+    @Override
+    public DnsQueryLifecycleObserver queryRedirected(List<InetSocketAddress> nameServers) {
+        registry.counter(meterIdPrefix.name().concat(".redirected"),
+                Arrays.asList(Tag.of(NAME_TAG, question.name()), Tag.of(SERVER_TAG,
+                        nameServers.stream().map(addr -> addr.getAddress().toString())
+                                .collect(Collectors.joining(","))))).increment();
+        return this;
+    }
+
+    @Override
+    public DnsQueryLifecycleObserver queryCNAMEd(DnsQuestion cnameQuestion) {
+        registry.counter(meterIdPrefix.name().concat(".cnamed"),
+                Arrays.asList(Tag.of(NAME_TAG, question.name()),
+                        Tag.of(CNAME_TAG, cnameQuestion.name()))).increment();
+        return this;
+    }
+
+    @Override
+    public DnsQueryLifecycleObserver queryNoAnswer(DnsResponseCode code) {
+        registry.counter(meterIdPrefix.name().concat(".noanswer"),
+                Arrays.asList(Tag.of(NAME_TAG, question.name()),
+                        Tag.of(CODE_TAG, code.toString()))).increment();
+        return this;
+    }
+
+    @Override
+    public void queryFailed(Throwable cause) {
+        registry.counter(meterIdPrefix.name(),
+                Arrays.asList(Tag.of(NAME_TAG, question.name()), Tag.of(RESULT_TAG, "failure"),
+                        Tag.of(CAUSE_TAG, cause.getMessage()))).increment();
```

Comment:
I'm somewhat concerned about using an exception message as a tag value because it can increase the cardinality dramatically. How about categorizing the exceptions into a few?

Reply:
Can you elaborate on this, please?  If I understand correctly we want to see what is the type of exception and use that to add CAUSE_TAG instead of cause.getMessage()?

Reply:
>  If I understand correctly we want to see what is the type of exception and use that to add CAUSE_TAG instead of cause.getMessage()?

Yes, if Netty gives different exception type for different causes. If not, we'll have to categorize all possible exception types and messages into a set of well-defined strings. Makes sense?

Reply:
Netty will return a runtimeexception of type DnsResolveContext::DnsResolveContextException which is RuntimeException. There are 5 different types of these exceptions defined in https://github.com/netty/netty/blob/fde6bc88859a41b2bdd5aae00beef09b1b17bda7/resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java#L65.  We also receive DnsNameResolverTimeoutException exception when there are time outs.

Reply:
I see. Thanks for looking into. We need to do some text pattern matching in this case. You might want to contribute to Netty so that each case is distinguishable by exception type later, though.